domain/src/main/java/org/oppia/android/domain/onboarding/DeprecationController.kt

```diff
@@ -0,0 +1,143 @@
+    cacheStoreFactory.create(
+      "deprecation_store",
+      DeprecationResponseDatabase.getDefaultInstance()
+    )
+  }
+
+  /**
+   * Enum states for the possible outcomes of a deprecation action.
+   */
+  private enum class DeprecationActionStatus {
+    /** Indicates that the deprecation operation succeeded. */
+    SUCCESS,
+
+    /** Indicates that a deprecation write operation failed. */
+    FAILED_TO_STORE_DEPRECATION_RESPONSE,
+
+    /**
+     * Indicates that a deprecation read operation failed. This is usually the result when a
+     * requested [DeprecationResponse] was not found.
+     */
+    DEPRECATION_RESPONSE_NOT_FOUND
+  }
+
+  init {
+    /**
+     * Prime the cache ahead of time so that the deprecation response can be retrieved
+     * synchronously.
+     */
+    deprecationStore.primeInMemoryAndDiskCacheAsync(
+      updateMode = PersistentCacheStore.UpdateMode.UPDATE_ALWAYS,
+      publishMode = PersistentCacheStore.PublishMode.PUBLISH_TO_IN_MEMORY_CACHE
+    ) {
+      it.toBuilder().build()
+    }.invokeOnCompletion { primeFailure ->
+      primeFailure?.let {
+        oppiaLogger.e(
+          "DeprecationController",
+          "Failed to prime cache ahead of data retrieval for DeprecationController.",
+          primeFailure
+        )
+      }
+    }
+  }
+
+  private val deprecationDataProvider by lazy { computeDeprecationProvider() }
+
+  private fun computeDeprecationProvider(): DataProvider<DeprecationResponseDatabase> {
+    return deprecationStore.transform(DEPRECATION_PROVIDER_ID) { deprecationResponsesDatabase ->
+      DeprecationResponseDatabase.newBuilder().apply {
+        appDeprecationResponse = deprecationResponsesDatabase.appDeprecationResponse
+        osDeprecationResponse = deprecationResponsesDatabase.osDeprecationResponse
+      }.build()
+    }
+  }
+
+  /**
+   * Returns a [DataProvider] containing the the [DeprecationResponseDatabase], which in turn
+   * affects what initial app flow the user is directed to.
+   */
+  fun getDeprecationDatabase(): DataProvider<DeprecationResponseDatabase> = deprecationDataProvider
+
+  /**
+   * Stores a new [DeprecationResponse] to the cache.
+   *
+   * @param deprecationResponse the deprecation response to be stored.
+   * @return [AsyncResult] of the deprecation action.
+   */
+  fun setDeprecationResponse(deprecationResponse: DeprecationResponse): DataProvider<Any?> {
+    val deferred = deprecationStore.storeDataWithCustomChannelAsync(
+      updateInMemoryCache = true
+    ) {
+      val deprecationBuilder = if (deprecationResponse.deprecationNoticeType
+        == DeprecationNoticeType.APP_DEPRECATION
+      ) {
+        it.toBuilder()
+          .setAppDeprecationResponse(deprecationResponse)
+          .build()
+      } else {
+        it.toBuilder()
+          .setOsDeprecationResponse(deprecationResponse)
+          .build()
+      }
+
+      Pair(deprecationBuilder, DeprecationActionStatus.SUCCESS)
+    }
```

Comment:
```suggestion
      { deprecationResponseDb ->
        val deprecationBuilder = deprecationResponseDb.toBuilder().apply {
          if (deprecationResponse.deprecationNoticeType == DeprecationNoticeType.APP_DEPRECATION)
            setAppDeprecationResponse(deprecationResponse)
          else
            setOsDeprecationResponse(deprecationResponse)
        }
          .build()

        Pair(deprecationBuilder, DeprecationActionStatus.SUCCESS)
      }
```

This is a little cleaner as it avoids repetition of `  it.toBuilder()` and `build()`

Reply:
I have implemented this change.